### PR TITLE
Add stack traces to logs

### LINKED ISSUE #1165: Stacktraces in logs
Every now and then people come in on Discord asking about crashes, especially with custom levels. Typically it's natural to refer to the the logs for crash details in these situations. However, our current builds don't provide that information and cut off the logs instead, rendering them meaningless for debugging purposes. It would be helpful to implement meaningful stacktrace reporting in the logs when crashes happen.

### PR DESCRIPTION
**Pull request by rr-**
#### Checklist

- [X] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [X] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description

Resolves #1165

### CODE REVIEW

**Comment by rr-** [comment]
@walkawayy can you please check which gives better stacktraces on Windows, the latest commit, or the one without libbacktrace (eg second to last)? On wine, my stacktrace using libstacktrace is more detailed, but it only displays 2-3 entries for an example 1/0 division I added, while with the WinAPI approach I'm getting it in full (although it's not able to resolve every symbol).

Using POSIX signals, while supported by mingw64, suffers from a similar problem and as such I can't see this solution using only generic code.

**Comment by walkawayy** [comment]
All 3 commits give me this error in WSL Windows 11:
```
Library opengl32 found: YES
Found pkg-config: /usr/bin/i686-w64-mingw32-pkg-config (0.29.2)
Run-time dependency libavcodec found: YES 58.134.100
Run-time dependency libavformat found: YES 58.76.100
Run-time dependency libavutil found: YES 56.70.100
Found CMake: NO
Run-time dependency libbacktrace found: NO (tried pkgconfig)

meson.build:38:16: ERROR: Dependency "libbacktrace" not found, tried pkgconfig
```

**Comment by rr-** [comment]
Hey @walkawayy since we're changing the Dockerfile and adding a new dependency, you'll need to run `make build-docker-image` before running a regular build.

**Comment by walkawayy** [comment]
Latest commit gave like no info on the crash (I screwed with the text in the option graphics menu):
```
../../src/specific/s_log_windows.c 59 S_Log_LogStackTraces == CRASH REPORT ==
../../src/specific/s_log_windows.c 60 S_Log_LogStackTraces STACK TRACE:
../../src/specific/s_log_windows.c 33 S_Log_FullTrace PC: 6fb99c, FILE: ../../src/specific/s_log_windows.c, LINE: 63, FUNC: S_Log_LogStackTraces
../../src/specific/s_log_windows.c 33 S_Log_FullTrace PC: 6fb9c4, FILE: ../../src/specific/s_log_windows.c, LINE: 70, FUNC: S_Log_CrashHandler@4
../../src/specific/s_log_windows.c 33 S_Log_FullTrace PC: 752f2d9f, FILE: ??, LINE: 0, FUNC: ??
```

Commit `609e996b2d5a76cfe1bfb5c889f98b5d9e9bf5ae` gave me a ton of build errors, like 200 lines of this:
```
FAILED: TR1X.exe.p/src_specific_s_log_windows.c.obj 
/usr/bin/i686-w64-mingw32-gcc -ITR1X.exe.p -I. -I../.. -I../../src -I/ext//include -I/ext/include -I/ext/include/SDL2 -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -std=c11 -O0 -g -Wno-unused -D_GNU_SOURCE -DMESON_BUILD -Dmain=SDL_main -MD -MQ TR1X.exe.p/src_specific_s_log_windows.c.obj -MF TR1X.exe.p/src_specific_s_log_windows.c.obj.d -o TR1X.exe.p/src_specific_s_log_windows.c.obj -c ../../src/specific/s_log_windows.c
In file included from /usr/share/mingw-w64/include/dbghelp.h:17,
                 from ../../src/specific/s_log_windows.c:5:
/usr/share/mingw-w64/include/psdk_inc/_dbg_LOAD_IMAGE.h:24:5: error: unknown type name ‘PSTR’
   24 |     PSTR ModuleName;
      |     ^~~~
/usr/share/mingw-w64/include/psdk_inc/_dbg_LOAD_IMAGE.h:25:5: error: unknown type name ‘HANDLE’
   25 |     HANDLE hFile;
      |     ^~~~~~
/usr/share/mingw-w64/include/psdk_inc/_dbg_LOAD_IMAGE.h:26:5: error: unknown type name ‘PUCHAR’
   26 |     PUCHAR MappedAddress;
      |     ^~~~~~
/usr/share/mingw-w64/include/psdk_inc/_dbg_LOAD_IMAGE.h:30:5: error: unknown type name ‘PIMAGE_NT_HEADERS32’
   30 |     PIMAGE_NT_HEADERS32 FileHeader;
      |     ^~~~~~~~~~~~~~~~~~~
/usr/share/mingw-w64/include/psdk_inc/_dbg_LOAD_IMAGE.h:32:5: error: unknown type name ‘PIMAGE_SECTION_HEADER’
   32 |     PIMAGE_SECTION_HEADER LastRvaSection;
      |     ^~~~~~~~~~~~~~~~~~~~~
/usr/share/mingw-w64/include/psdk_inc/_dbg_common.h:376:5: error: unknown type name ‘DWORD’
  376 |     DWORD SizeOfStruct;
```

**Comment by rr-** [comment]
@walkawayy can you please try now?

**Comment by walkawayy** [comment]
Ok. Which commit? Both or just latest?

**Comment by walkawayy** [comment]
Latest commit:
```
../../src/specific/s_log_windows.c 37 S_Log_LogStackTraces == CRASH REPORT ==
../../src/specific/s_log_windows.c 109 S_Log_LogStackTraces 0x00000000: ???
../../src/specific/s_log_windows.c 106 S_Log_LogStackTraces 0x752F2C20: UnhandledExceptionFilter
../../src/specific/s_log_windows.c 106 S_Log_LogStackTraces 0x774DBCC0: RtlInitializeExceptionChain
../../src/specific/s_log_windows.c 106 S_Log_LogStackTraces 0x774DBBF0: RtlClearBits
```

I just added a `return` to the beginning of `static void Option_GraphicsInitText(void)`. It's an easy way to crash the game lol.

**Comment by rr-** [comment]
Ugh. On wine, I get:

```
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x0048BC4B: S_Log_CrashHandler (Z:\app\src\specific\s_log_windows.c:120)
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x7B644160: UnhandledExceptionFilter (Z:\build\wine\src\wine\dlls\kernelbase\debug.c:733)
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x7BD69350: __wine_exception_handler (Z:\build\wine\src\wine\dlls\winecrt0\exception.c:108)
../../src/specific/s_log_windows.c 109 S_Log_LogStackTraces 0x7BD69350: ???
../../src/specific/s_log_windows.c 109 S_Log_LogStackTraces 0x7BD69350: ???
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x7BD54BC0: dispatch_exception (Z:\build\wine\src\wine\dlls\ntdll\signal_i386.c:247)
../../src/specific/s_log_windows.c 109 S_Log_LogStackTraces 0x7BD54BC0: ???
../../src/specific/s_log_windows.c 109 S_Log_LogStackTraces 0x7BD54BC0: ???
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x00444454: Option_Graphics (Z:\app\src\game\option\option_graphics.c:515)
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x00441604: Option_DoInventory (Z:\app\src\game\option\option.c:37)
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x0044DEAE: Phase_Inventory_Control (Z:\app\src\game\phase\phase_inventory.c:883)
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x0044C494: Phase_Control (Z:\app\src\game\phase\phase.c:33)
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x0044C663: Phase_Run (Z:\app\src\game\phase\phase.c:127)
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x0040E340: Game_MainMenu (Z:\app\src\game\game\game_main_menu.c:10)
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x0045E167: Shell_Main (Z:\app\src\game\shell.c:249)
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x0048B34A: SDL_main (Z:\app\src\specific\s_shell.c:308)
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x00A36250: main_getcmdline (Z:\SDL\src\main\windows\SDL_windows_main.c:80)
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x00A36460: WinMain (Z:\SDL\src\main\windows\SDL_windows_main.c:110)
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x00C13DB0: main (Z:\home\rr-\work\priv\TR1X\test\build\i686-w64-mingw32-i686-w64-mingw32-crt\mingw-w64-crt\crt\crt0_c.c:20)
../../src/specific/s_log_windows.c 102 S_Log_LogStackTraces 0x00401170: __tmainCRTStartup (Z:\home\rr-\work\priv\TR1X\test\build\i686-w64-mingw32-i686-w64-mingw32-crt\mingw-w64-crt\crt\crtexe.c:335)
../../src/specific/s_log_windows.c 109 S_Log_LogStackTraces 0x00401170: ???
../../src/specific/s_log_windows.c 109 S_Log_LogStackTraces 0x00401170: ???
../../src/specific/s_log_windows.c 109 S_Log_LogStackTraces 0x00401170: ???
```

On Linux:

```
../../src/specific/s_log_linux.c 39 S_Log_SignalHandler == CRASH REPORT ==
../../src/specific/s_log_linux.c 40 S_Log_SignalHandler SIGNAL: 11
../../src/specific/s_log_linux.c 41 S_Log_SignalHandler STACK TRACE:
../../src/specific/s_log_linux.c 27 S_Log_FullTrace 0x5A671D91: S_Log_SignalHandler (../../src/specific/s_log_linux.c:44)
../../src/specific/s_log_linux.c 31 S_Log_FullTrace 0xA349676F: ???
../../src/specific/s_log_linux.c 27 S_Log_FullTrace 0x5A61D104: Option_GraphicsUpdateArrows (../../src/game/option/option_graphics.c:223)
../../src/specific/s_log_linux.c 27 S_Log_FullTrace 0x5A61DD7C: Option_Graphics (../../src/game/option/option_graphics.c:513)
../../src/specific/s_log_linux.c 27 S_Log_FullTrace 0x5A61A3C9: Option_DoInventory (../../src/game/option/option.c:36)
../../src/specific/s_log_linux.c 27 S_Log_FullTrace 0x5A6299BF: Phase_Inventory_Control (../../src/game/phase/phase_inventory.c:881)
../../src/specific/s_log_linux.c 27 S_Log_FullTrace 0x5A626A58: Phase_Control (../../src/game/phase/phase.c:33)
../../src/specific/s_log_linux.c 27 S_Log_FullTrace 0x5A626CBC: Phase_Run (../../src/game/phase/phase.c:127)
../../src/specific/s_log_linux.c 27 S_Log_FullTrace 0x5A5DECAA: Game_MainMenu (../../src/game/game/game_main_menu.c:9)
../../src/specific/s_log_linux.c 27 S_Log_FullTrace 0x5A63B47C: Shell_Main (../../src/game/shell.c:249)
../../src/specific/s_log_linux.c 27 S_Log_FullTrace 0x5A67180C: main (../../src/specific/s_shell.c:306)
../../src/specific/s_log_linux.c 31 S_Log_FullTrace 0xA347FCCF: ???
../../src/specific/s_log_linux.c 31 S_Log_FullTrace 0xA347FD89: ???
../../src/specific/s_log_linux.c 31 S_Log_FullTrace 0x5A5D0224: ???
../../src/specific/s_log_linux.c 31 S_Log_FullTrace 0xFFFFFFFF: ???
```

I think it's as good as we can get…

**Comment by walkawayy** [comment]
Ok no problem. It's better than what we had before.

File: meson.build [hunk 1]
```diff
@@ -304,6 +305,17 @@ dependencies = [
...
 
+if dep_backtrace.found() and host_machine.system() == 'linux'
+  sources += ['src/specific/s_log_linux.c']
+  dependencies += [dep_backtrace]
+elif dep_backtrace.found() and host_machine.system() == 'windows'
+  sources += ['src/specific/s_log_windows.c']
+  dep_win_dbghelp = c_compiler.find_library('dbghelp')
+  dependencies += [dep_win_dbghelp]
+else
+  sources += ['src/specific/s_log_unkown.c']
```
Comment (lahm86) [inline meson.build]: A little typo - `s_log_unknown.c`

Comment (rr-) [reply meson.build]: @lahm86 fixed

**Review by lahm86 — APPROVED** [review]

**Review by walkawayy — APPROVED** [review]

**Pull request merged**